### PR TITLE
Alt+Enter newline shortcut

### PR DESCRIPTION
**Pull request by goolord**
in most messaging apps, shift+enter enters a newline character. this pr adds that behavior.
currently i'm unable to test if this behaves as intended because running gurk panics for me

### CODE REVIEW

**Comment by boxdot** [comment]
Thank you! I had the same implementation but it does not work: https://github.com/crossterm-rs/crossterm/issues/400.

off-topic: If you have the line number where it panics, I might be able to fix it.

**Comment by goolord** [comment]
I'll create a seperate issue about the panic.
What about Alt+Enter? That's how my terminal allows entering newlines

**Comment by boxdot** [comment]
>  What about Alt+Enter? That's how my terminal allows entering newlines

Sounds good!

**Comment by goolord** [comment]
Ok, I write very little rust but hopefully nothing looks off. The solution is a little janky, crossterm reads Alt+Enter as `'\r'`, so all of the logic is actually in `draw_chat`. Even though this eliminated the need for the key modifiers, I left the `put_char` function factored out because I think it's nice.

**Comment by boxdot** [comment]
> Ok, I write very little rust but hopefully nothing looks off. 

No worries. :) I appreciate the contribution a lot.

> The solution is a little janky, crossterm reads Alt+Enter as '\r', so all of the logic is actually in draw_chat. Even though this eliminated the need for the key modifiers, I left the put_char function factored out because I think it's nice.

I like the factored out function. I was thinking that now, in the current implementation, we send `\r` to the Signal server. I guess we could replace `\r` just by `\n` in the input field on input? Then there is no need for special handling in ui, right?

**Comment by goolord** [comment]
I'm not familiar enough with `tui` to know why this is, but the `'\n'` (and `'\r'`) character doesn't actually visually create a newline, instead each index in `lines: Vec<String>` is rendered as a new line.. it ends up being fine I think because it helps in calculating the y position of the input cursor. Also, I just pushed a commit which fixes a few bugs that were introduced by rendering '\n' characters

File: src/ui.rs [hunk 1]
```diff
@@ -56,18 +56,35 @@ pub fn draw<B: Backend>(f: &mut Frame<B>, app: &mut App) {
...
+            .input
+            .chars()
+            .enumerate()
+            .fold(Vec::new(), |mut lines, (idx, c)| {
+                if idx % text_width == 0 {
+                    lines.push(String::new());
+                }
+                match c {
+                    '\n' => {
+                        lines.last_mut().unwrap().push('\n');
```
Comment (boxdot) [inline src/ui.rs:70]: I don't think this one is needed.

File: src/ui.rs [hunk 2]
```diff
@@ -56,18 +56,35 @@ pub fn draw<B: Backend>(f: &mut Frame<B>, app: &mut App) {
...
+    let lines: Vec<String> =
+        app.data
+            .input
+            .chars()
+            .enumerate()
+            .fold(Vec::new(), |mut lines, (idx, c)| {
+                if idx % text_width == 0 {
+                    lines.push(String::new());
+                }
+                match c {
```
Comment (boxdot) [inline src/ui.rs:68]: You could combine this match with the `if` above: `if idx % text_width == 0 || c == '\n'`.

File: src/ui.rs [hunk 3]
```diff
@@ -56,18 +56,35 @@ pub fn draw<B: Backend>(f: &mut Frame<B>, app: &mut App) {
...
+                    '\n' => {
+                        lines.last_mut().unwrap().push('\n');
+                        lines.push(String::new())
+                    }
+                    _ => lines.last_mut().unwrap().push(c),
+                }
+                lines
+            });
+    // chars since newline on `cursor_y` line
```
Comment (boxdot) [inline src/ui.rs:77]: @goolord: Ah, I see. The input handling is a little bit hacky. I think I have a solution though. You can calculate here:

```rust
let last_line_width = lines.last().map(|l| l.width()).unwrap_or(0);
let num_input_lines = lines.len().max(1);
```

Then add `last_line_width + 1` to `x`, and `num_input_lines` to `y` in lines 114, 116.

Comment (goolord) [reply src/ui.rs:77]: the cursor isn't always on the last line, so this code calculates which line its on and the offset from the beginning of that line

File: src/ui.rs (hunk 2 above)
Comment (goolord) [reply src/ui.rs:68]: that's subtly different, in the case that `'\n'` is input at the end of a line that will push a new string and then draw the newline character, which will mess up the cursor positioning and rendering in weird ways

**Review by boxdot — APPROVED** [review]
Private life got me, new job, etc... :umbrella: I am really sorry for being not responsive. I'll try to be faster from now on.

**Comment by boxdot** [comment]
@goolord I hope you are fine with me changing this behavior. With #131, Alt+Enter will be remapped to switch between the multi-line and single-line modes.

**Pull request merged**